File: tests/Integration/StorageAdmission/sdk/worker.py
import asyncio
from pathlib import Path

from durable_workflow import Client, Worker, activity, workflow


@activity.defn(name="storage.python.echo")
async def echo() -> str:
    value = "python:" + "x" * 262144
    with Path("/observation/python.effects").open("a") as output:
        output.write("executed\n")
    async with asyncio.timeout(120):
        while not Path("/observation/release-activities").is_file():
            await asyncio.sleep(0.1)
    Path("/observation/python.returned").write_text("ready")
    return value


@workflow.defn(name="storage.python")
class StorageWorkflow:
    def run(self, context):
        import hashlib

        value = yield context.schedule_activity(
            "storage.python.echo", [], start_to_close_timeout=180
        )
        return {"runtime": "python", "bytes": len(value), "sha256": hashlib.sha256(value.encode()).hexdigest()}


async def main():
    async with Client("http://server:8080", token="storage-fixture") as client:
        await Worker(
            client, task_queue="storage-python", worker_id="storage-python",
            workflows=[StorageWorkflow], activities=[echo],
            poll_timeout=1, heartbeat_interval=5,
        ).run()


asyncio.run(main())

File: tests/Integration/StorageAdmission/sdk/worker.rs
use std::{fs, io::Write, time::Duration};

use durable_workflow::{Client, Result, Worker};
use serde_json::json;
use sha2::{Digest, Sha256};

#[tokio::main]
async fn main() -> Result<()> {
    let client = Client::builder("http://server:8080")
        .token(Some("storage-fixture".to_string()))
        .build()?;
    let mut worker = Worker::new(client, "storage-rust")
        .worker_id("storage-rust")
        .poll_timeout(Duration::from_secs(1));
    worker.register_workflow("storage.rust", |context, _input| async move {
        let value = context.activity("storage.rust.echo", json!([])).await?;
        let text = value.as_str().expect("fixture activity returns a string");
        Ok(json!({"runtime": "rust", "bytes": text.len(), "sha256": format!("{:x}", Sha256::digest(text.as_bytes()))}))
    });
    worker.register_activity("storage.rust.echo", |_context, _input| async move {
        let value = format!("rust:{}", "x".repeat(262144));
        let mut effects = fs::OpenOptions::new().create(true).append(true)
            .open("/observation/rust.effects").expect("fixture effects file");
        writeln!(effects, "executed").expect("record fixture effect");
        tokio::time::timeout(Duration::from_secs(120), async {
            while !std::path::Path::new("/observation/release-activities").is_file() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }).await.expect("fixture release barrier timed out");
        fs::write("/observation/rust.returned", "ready").expect("fixture return marker");
        Ok(json!(value))
    });
    worker.run().await
}
